### PR TITLE
Sid's work on debugging

### PR DESCRIPTION
**Pull request by manzoid**

### CODE REVIEW

File: modules/debugging/overview.md [hunk 1]
```diff
@@ -1,23 +1,229 @@
...
+Description:
+These are usually very easy to spot but you are consistantly off by one in your 
+measure or test in something.
+
+Note:
+Generally there is an edge case that is missed or an index that is not being accounted for.
+If there is a patter of always being off by one in your result that means you are probably missing the first
+or last case.
+
+### Incorrect level of Nesting
```
Comment (manzoid) [inline modules/debugging/overview.md:160]: i think this is more generally referring to nested logic rather than just processing a matrix via nested loops.

And there is some piece of logic being at the wrong level of that nesting.

Nesting doesn't just happen in loops.  E.g., it can happen if there are nested conditionals, or across functions, if there are helper functions being called.

File: modules/debugging/overview.md [hunk 2]
```diff
@@ -1,23 +1,229 @@
...
+
+Ways to tell this is the problem:
+1. Type errors usually inform you of this.
+2. Data has not been changed from original form.
+
+Note:
+Make sure you reorder your input to make sure if fits what you are excpecting before
+you proceed with working with it. 
+
+<!-- ### Incorrect order of Results ISNT THIS SORT OF AN OBVIOUS ERROR THAT DOES'T REALLY REQUIRE CHANGE-->
```
Comment (manzoid) [inline modules/debugging/overview.md:148]: not sure what you mean.  If you return an array that contains the results not in the specified order, then it's a bug.

There are languages such as Go that allow you return more than one value.  If those values are both of the same type but the order is reversed, then you have the same problem as a function that has two parameters of the same type and the order is reversed at the call site.

File: modules/debugging/overview.md [hunk 3]
```diff
@@ -1,23 +1,229 @@
...
+
+
+
+<!-- ### Approach 6: Keep an audit trail. NOT SURE HOW TO MAKE THIS AN APPROACH AGAIN FEEL LIKE ITS MORE OF A PRACTICE
+FOLLOWING WHAT YOURE DOING CLOSESLY
+### Approach 7: Check the plug. NOT REALLY A PROCESS BUT GIVING IT A BREAK IS MORE OF A PRACTICE
+
+1. Get a fresh view.
+
+### approach: if you didn't fix it, it ain't fixed.  YEAH BUT HOW DOES IT HELP-->
```
Comment (manzoid) [inline modules/debugging/overview.md:59]: The linked debugging book that we are using for inspiration is very short, so to answer a question like this, skimming the corresponding section is a good first step.

Anyway, this one is saying that you aren't done until you've finished verifying it fully. So it's part of the debugging process in the sense of being a final step.

File: modules/debugging/overview.md [hunk 4]
```diff
@@ -1,23 +1,229 @@
...
+### Approach 4: Divide and conquer.
+
+If you divide your code into multiple parts and make sure each works. Together
+the usually will as well. 
+
+1. Divide code into large parts.
+2. Make the parts smaller and smaller until the error is on one small portion.
+3. Use any other approach to fix this error. 
+
+<!-- ### Approach 5: Change one thing at a time.  FEEL LIKE THIS IS SORT OF A BAD DEBUGGING PROCESS ON ITS OWN-->
```
Comment (manzoid) [inline modules/debugging/overview.md:49]: Not sure what you mean.  First, I don't think the idea is that each of the bullet points is meant to represent an entire process that's used in isolation.

And changing one thing at a time is part of the essence of the scientific method, wouldn't you say.  To establish or at least increase the probability of clear causation, isolate one variable.

File: modules/debugging/overview.md [hunk 5]
```diff
@@ -1,23 +1,229 @@
...
+### Incorrect level of Nesting 
+Description:
+This usually means in a nxn object exactly an (n-1)x(n-1) amount of data is being changed or 
+used. 
+
+Ways to tell:
+1. Only a part of your data (usually the first row or column) has been changed.
+
+
+<!-- ### Misnamed property DONT THINK THIS IS NECESSARY PERSONALLY-->
```
Comment (manzoid) [inline modules/debugging/overview.md:169]: Why? This is a very common bug, especially in scripting languages.

File: modules/debugging/overview.md [hunk 6]
```diff
@@ -1,23 +1,229 @@
...
+1. Your math doesn't output the right number.
+
+Note: This can be from:
+1. Confusing median for mean. 
+2. Divide by zero.
+3. Error on negative numbers, or 0.
+
+<!-- ### typo or reversal in translating a formula NOT SURE I FOLLOW THIS ONE-->
+
+<!-- ### incorrectly consuming an api response object WON'T YOUR FETCH/GET REQUEST TELL YOU WHAT YOU GET AND
```
Comment (manzoid) [inline modules/debugging/overview.md:186]: Assuming that the response comes in as a JSON object, first there is no type information in JSON, second, by default nobody knows what is "correct" because it's just some arbitrary data-blob.  Correctness is a concern of the problem domain, which a language or primitive types don't really help with.

File: modules/debugging/overview.md [hunk 7]
```diff
@@ -1,23 +1,229 @@
...
+
+Ways to tell:
+1. Your math doesn't output the right number.
+
+Note: This can be from:
+1. Confusing median for mean. 
+2. Divide by zero.
+3. Error on negative numbers, or 0.
+
+<!-- ### typo or reversal in translating a formula NOT SURE I FOLLOW THIS ONE-->
```
Comment (manzoid) [inline modules/debugging/overview.md:184]: Sometimes we are implementing from a whitepaper or, more commonly, somebody's specification doc that includes some kind of calculations.  We have to do the translation of the formulas -- even if they are just arithmetic or minor algebra --correctly, or we have introduced a bug.  So when such a bug emerges, we have to look closely at the original specification and the accompanying correct samples, which hopefully have been turned properly into unit tests, in order to determine the error.

**Comment by manzoid** [comment]
@sidarth-raman I think focusing on making the actual exercises might be the most productive use of time.  It's the most concrete way too to make sure we are on the same page about what each of those bullet points is meant to represent.  If you find yourself having difficulty coming up with a suitable small exercise then let's discuss. 

By the way these exercises should live in a separate repo or repos, similar to how the Functions I module is starting to have separate exercise repos such as https://github.com/embarkpark/functions_avoiding_side_effects
